discovery_engine_core/core/src/stack/filters.rs

```diff
@@ -39,52 +36,57 @@ impl ArticleFilter for DuplicateFilter {
     fn apply(
         history: &[HistoricDocument],
         stack: &[Document],
-        articles: Vec<Article>,
+        mut articles: Vec<Article>,
     ) -> Result<Vec<Article>, GenericError> {
-        let urls = history
+        let urls = stack
             .iter()
-            .map(|doc| doc.url.as_str())
-            .chain(stack.iter().map(|doc| doc.resource.url.as_str()))
-            .collect::<HashSet<_>>();
-
-        let titles = history
+            .map(|doc| (doc.resource.url.as_str(), doc.resource.rank))
+            .chain(history.iter().map(|doc| (doc.url.as_str(), 0)))
+            .fold(HashMap::new(), |mut urls, (url, rank)| {
+                let best_rank = urls.entry(url).or_insert(rank);
+                if rank < *best_rank {
+                    *best_rank = rank;
+                };
```

Comment:
🟢  we could write this a bit more consise such that there is no additional check done in the or-case. similarly for the `titles`.
```suggestion
                urls.entry(url)
                    .and_modify(|best_rank| {
                        if rank < *best_rank {
                            *best_rank = rank;
                        }
                    })
                    .or_insert(rank);
```